File: sdk/php/src/Interchange/Importer.php
<?php

declare(strict_types=1);

namespace AnyVali\Interchange;

use AnyVali\AnyValiDocument;
use AnyVali\Schema;
use AnyVali\UnknownKeyMode;
use AnyVali\Schemas\{
    AnySchema,
    ArraySchema,
    BoolSchema,
    EnumSchema,
    IntSchema,
    IntersectionSchema,
    LiteralSchema,
    NeverSchema,
    NullSchema,
    NullableSchema,
    NumberSchema,
    ObjectSchema,
    OptionalSchema,
    RecordSchema,
    RefSchema,
    StringSchema,
    TupleSchema,
    UnionSchema,
    UnknownSchema,
};

final class Importer
{
    /** @var array<string, Schema> */
    private array $resolved = [];
    /** @var array<string, bool> */
    private array $active = [];
    /** @var array<string, list<RefSchema>> */
    private array $pending = [];

    private function __construct()
    {
    }

    /**
     * Import a schema from an AnyValiDocument.
     */
    public static function import(AnyValiDocument $doc): Schema
    {
        return self::importNode($doc->root, $doc->definitions);
    }

    /**
     * Import from a JSON string.
     */
    public static function importJson(string $json): Schema
    {
        return self::import(AnyValiDocument::fromJson($json));
    }

    /**
     * Import from array.
     * @param array<string, mixed> $data
     */
    public static function importArray(array $data): Schema
    {
        return self::import(AnyValiDocument::fromArray($data));
    }

    /**
     * Import a single schema node.
     * @param array<string, mixed> $node
     * @param array<string, array<string, mixed>> $definitions
     */
    public static function importNode(array $node, array $definitions = []): Schema
    {
        return (new self())->node($node, $definitions);
    }

    private function node(array $node, array $definitions): Schema
    {
        $kind = $node['kind'] ?? null;

        if ($kind === null) {
            throw new \RuntimeException('Schema node missing "kind" field');
        }

        $schema = match ($kind) {
            'string' => $this->importString($node),
            'number', 'float64', 'float32' => $this->importNumber($node, $kind),
            'int', 'int8', 'int16', 'int32', 'int64',
            'uint8', 'uint16', 'uint32', 'uint64' => $this->importInt($node, $kind),
            'bool' => $this->importBool($node),
            'null' => new NullSchema(),
            'any' => new AnySchema(),
            'unknown' => new UnknownSchema(),
            'never' => new NeverSchema(),
            'literal' => new LiteralSchema($node['value'] ?? null),
            'enum' => new EnumSchema($node['values'] ?? []),
            'array' => $this->importArray2($node, $definitions),
            'tuple' => $this->importTuple($node, $definitions),
            'object' => $this->importObject($node, $definitions),
            'record' => $this->importRecord($node, $definitions),
            'union' => $this->importUnion($node, $definitions),
            'intersection' => $this->importIntersection($node, $definitions),
            'optional' => $this->importOptional($node, $definitions),
            'nullable' => $this->importNullable($node, $definitions),
            'ref' => new RefSchema($node['ref'] ?? ''),
            default => throw new \RuntimeException("Unsupported schema kind: {$kind}"),
        };

        // Apply coerce
        if (isset($node['coerce'])) {
            $schema = $schema->coerce($node['coerce']);
        }

        // Apply default
        if (array_key_exists('default', $node)) {
            $schema = $schema->default($node['default']);
        }

        if ($schema instanceof RefSchema) {
            $this->resolveRef($schema, $definitions);
        }
        return $schema;
    }

    /**
     * @param array<string, mixed> $node
     */
    private function importString(array $node): StringSchema
    {
        $schema = new StringSchema();
        if (isset($node['minLength'])) $schema = $schema->minLength((int)$node['minLength']);
        if (isset($node['maxLength'])) $schema = $schema->maxLength((int)$node['maxLength']);
        if (isset($node['pattern'])) $schema = $schema->pattern($node['pattern']);
        if (isset($node['startsWith'])) $schema = $schema->startsWith($node['startsWith']);
        if (isset($node['endsWith'])) $schema = $schema->endsWith($node['endsWith']);
        if (isset($node['includes'])) $schema = $schema->includes($node['includes']);
        if (isset($node['format'])) $schema = $schema->format($node['format']);
        return $schema;
    }

    /**
     * @param array<string, mixed> $node
     */
    private function importNumber(array $node, string $kind): NumberSchema
    {
        $schema = new NumberSchema($kind);
        if (isset($node['min'])) $schema = $schema->min($node['min']);
        if (isset($node['max'])) $schema = $schema->max($node['max']);
        if (isset($node['exclusiveMin'])) $schema = $schema->exclusiveMin($node['exclusiveMin']);
        if (isset($node['exclusiveMax'])) $schema = $schema->exclusiveMax($node['exclusiveMax']);
        if (isset($node['multipleOf'])) $schema = $schema->multipleOf($node['multipleOf']);
        return $schema;
    }

    /**
     * @param array<string, mixed> $node
     */
    private function importInt(array $node, string $kind): IntSchema
    {
        $schema = new IntSchema($kind);
        if (isset($node['min'])) $schema = $schema->min($node['min']);
        if (isset($node['max'])) $schema = $schema->max($node['max']);
        if (isset($node['exclusiveMin'])) $schema = $schema->exclusiveMin($node['exclusiveMin']);
        if (isset($node['exclusiveMax'])) $schema = $schema->exclusiveMax($node['exclusiveMax']);
        if (isset($node['multipleOf'])) $schema = $schema->multipleOf($node['multipleOf']);
        return $schema;
    }

    /**
     * @param array<string, mixed> $node
     */
    private function importBool(array $node): BoolSchema
    {
        return new BoolSchema();
    }

    /**
     * @param array<string, mixed> $node
     * @param array<string, array<string, mixed>> $definitions
     */
    private function importArray2(array $node, array $definitions): ArraySchema
    {
        // Accept "items", "item", and "array.items" keys for compatibility
        $itemsNode = $node['items'] ?? $node['item'] ?? $node['array.items'] ?? ['kind' => 'any'];
        $items = $this->node($itemsNode, $definitions);
        $schema = new ArraySchema($items);
        if (isset($node['minItems'])) $schema = $schema->minItems((int)$node['minItems']);
        if (isset($node['maxItems'])) $schema = $schema->maxItems((int)$node['maxItems']);
        return $schema;
    }

    /**
     * @param array<string, mixed> $node
     * @param array<string, array<string, mixed>> $definitions
     */
    private function importTuple(array $node, array $definitions): TupleSchema
    {
        $elements = array_map(
            fn(array $el) => $this->node($el, $definitions),
            $node['elements'] ?? [],
        );
        return new TupleSchema($elements);
    }

    /**
     * @param array<string, mixed> $node
     * @param array<string, array<string, mixed>> $definitions
     */
    private function importObject(array $node, array $definitions): ObjectSchema
    {
        $properties = [];
        foreach (($node['properties'] ?? []) as $key => $propNode) {
            $properties[$key] = $this->node($propNode, $definitions);
        }

        $required = $node['required'] ?? [];
        $unknownKeys = UnknownKeyMode::tryFrom($node['unknownKeys'] ?? 'strip')
            ?? UnknownKeyMode::Strip;

        return new ObjectSchema($properties, $required, $unknownKeys, true);
    }

    /**
     * @param array<string, mixed> $node
     * @param array<string, array<string, mixed>> $definitions
     */
    private function importRecord(array $node, array $definitions): RecordSchema
    {
        $valueNode = array_key_exists('values', $node) ? $node['values'] : ($node['valueSchema'] ?? null);
        if (!is_array($valueNode)) throw new \RuntimeException('Record schema missing or invalid "values"');
        $values = $this->node($valueNode, $definitions);
        return new RecordSchema($values);
    }

    /**
     * @param array<string, mixed> $node
     * @param array<string, array<string, mixed>> $definitions
     */
    private function importUnion(array $node, array $definitions): UnionSchema
    {
        // Accept "variants", "schemas", and "union.variants" keys for compatibility
        $variantsData = $node['variants'] ?? $node['schemas'] ?? $node['union.variants'] ?? [];
        $variants = array_map(
            fn(array $v) => $this->node($v, $definitions),
            $variantsData,
        );
        return new UnionSchema($variants);
    }

    /**
     * @param array<string, mixed> $node
     * @param array<string, array<string, mixed>> $definitions
     */
    private function importIntersection(array $node, array $definitions): IntersectionSchema
    {
        $allOf = array_map(
            fn(array $s) => $this->node($s, $definitions),
            $node['allOf'] ?? [],
        );
        return new IntersectionSchema($allOf);
    }

    /**
     * @param array<string, mixed> $node
     * @param array<string, array<string, mixed>> $definitions
     */
    private function importOptional(array $node, array $definitions): OptionalSchema
    {
        $inner = $this->node($node['schema'] ?? ['kind' => 'any'], $definitions);
        $schema = new OptionalSchema($inner);
        return $schema;
    }

    /**
     * @param array<string, mixed> $node
     * @param array<string, array<string, mixed>> $definitions
     */
    private function importNullable(array $node, array $definitions): NullableSchema
    {
        $inner = $this->node($node['schema'] ?? ['kind' => 'any'], $definitions);
        $schema = new NullableSchema($inner);
        return $schema;
    }

    /**
     * @param array<string, array<string, mixed>> $definitions
     */
    private function resolveRef(RefSchema $schema, array $definitions): void
    {
        $ref = $schema->getRef();
        $prefix = '#/definitions/';
        if (!str_starts_with($ref, $prefix)) return;
        $name = substr($ref, strlen($prefix));
        if (!isset($definitions[$name])) return;
        if (isset($this->resolved[$name])) {
            $schema->resolve($this->resolved[$name]);
            return;
        }
        if (isset($this->active[$name])) {
            $this->pending[$name][] = $schema;
            return;
        }
        $this->active[$name] = true;
        $this->pending[$name] = [$schema];
        try {
            $target = $this->node($definitions[$name], $definitions);
            $this->resolved[$name] = $target;
            foreach ($this->pending[$name] as $pending) {
                $pending->resolve($target);
            }
        } finally {
            unset($this->active[$name], $this->pending[$name]);
        }
    }
}

File: sdk/csharp/src/AnyVali/Interchange/Importer.cs
using AnyVali.Parse;
using AnyVali.Schemas;

namespace AnyVali.Interchange;

/// <summary>
/// Import an AnyValiDocument back into a live Schema.
/// </summary>
public static class Importer
{
    public static Schema ImportSchema(AnyValiDocument doc)
    {
        var definitions = doc.Definitions ?? new Dictionary<string, object?>();
        var resolvedDefs = new Dictionary<string, Schema>();

        Schema ImportNode(object? nodeObj)
        {
            var node = ToDict(nodeObj);
            var kind = node.TryGetValue("kind", out var k) ? k?.ToString() : null;

            Schema schema;

            switch (kind)
            {
                case "string":
                {
                    var s = new StringSchema();
                    if (node.TryGetValue("minLength", out var ml) && ml is not null)
                        s = s.MinLength(ToInt(ml));
                    if (node.TryGetValue("maxLength", out var mxl) && mxl is not null)
                        s = s.MaxLength(ToInt(mxl));
                    if (node.TryGetValue("pattern", out var p) && p is string ps)
                        s = s.Pattern(ps);
                    if (node.TryGetValue("startsWith", out var sw) && sw is string sws)
                        s = s.StartsWith(sws);
                    if (node.TryGetValue("endsWith", out var ew) && ew is string ews)
                        s = s.EndsWith(ews);
                    if (node.TryGetValue("includes", out var inc) && inc is string incs)
                        s = s.Includes(incs);
                    if (node.TryGetValue("format", out var f) && f is string fs)
                        s = s.Format(fs);
                    schema = s;
                    break;
                }
                case "number":
                case "float64":
                {
                    var s = kind == "float64" ? (NumberSchema)new Float64Schema() : new NumberSchema();
                    schema = ApplyNumericConstraints(s, node);
                    break;
                }
                case "float32":
                    schema = ApplyNumericConstraints(new Float32Schema(), node);
                    break;
                case "int":
                case "int64":
                {
                    var s = kind == "int64" ? (IntSchema)new Int64Schema() : new IntSchema();
                    schema = ApplyNumericConstraints(s, node);
                    break;
                }
                case "int8":
                    schema = ApplyNumericConstraints(new Int8Schema(), node);
                    break;
                case "int16":
                    schema = ApplyNumericConstraints(new Int16Schema(), node);
                    break;
                case "int32":
                    schema = ApplyNumericConstraints(new Int32Schema(), node);
                    break;
                case "uint8":
                    schema = ApplyNumericConstraints(new Uint8Schema(), node);
                    break;
                case "uint16":
                    schema = ApplyNumericConstraints(new Uint16Schema(), node);
                    break;
                case "uint32":
                    schema = ApplyNumericConstraints(new Uint32Schema(), node);
                    break;
                case "uint64":
                    schema = ApplyNumericConstraints(new Uint64Schema(), node);
                    break;
                case "bool":
                    schema = new BoolSchema();
                    break;
                case "null":
                    schema = new NullSchema();
                    break;
                case "any":
                    schema = new AnySchema();
                    break;
                case "unknown":
                    schema = new UnknownSchema();
                    break;
                case "never":
                    schema = new NeverSchema();
                    break;
                case "literal":
                {
                    var value = node.TryGetValue("value", out var v) ? v : null;
                    schema = new LiteralSchema(value);
                    break;
                }
                case "enum":
                {
                    var values = node.TryGetValue("values", out var v) && v is List<object?> list
                        ? list.Where(x => x is not null).Cast<object>().ToArray()
                        : Array.Empty<object>();
                    schema = new EnumSchema(values);
                    break;
                }
                case "array":
                {
                    var items = node.TryGetValue("items", out var it) ? ImportNode(it) : new AnySchema();
                    var s = new ArraySchema(items);
                    if (node.TryGetValue("minItems", out var mi) && mi is not null)
                        s = s.MinItems(ToInt(mi));
                    if (node.TryGetValue("maxItems", out var mxi) && mxi is not null)
                        s = s.MaxItems(ToInt(mxi));
                    schema = s;
                    break;
                }
                case "tuple":
                {
                    // Corpus uses "elements", export uses "items"
                    var elements = node.TryGetValue("elements", out var el) ? el
                        : node.TryGetValue("items", out var it) ? it : null;
                    var items = elements is List<object?> list
                        ? list.Select(ImportNode).ToArray()
                        : Array.Empty<Schema>();
                    schema = new TupleSchema(items);
                    break;
                }
                case "object":
                {
                    var shape = new Dictionary<string, Schema>();
                    var requiredSet = new HashSet<string>();
                    if (node.TryGetValue("required", out var req) && req is List<object?> reqList)
                    {
                        foreach (var r in reqList)
                            if (r is string rs) requiredSet.Add(rs);
                    }
                    if (node.TryGetValue("properties", out var props) && props is Dictionary<string, object?> propsDict)
                    {
                        foreach (var (propKey, propNode) in propsDict)
                        {
                            var propSchema = ImportNode(propNode);
                            if (!requiredSet.Contains(propKey))
                                propSchema = new OptionalSchema(propSchema);
                            shape[propKey] = propSchema;
                        }
                    }
                    var ukMode = UnknownKeyMode.Reject;
                    if (node.TryGetValue("unknownKeys", out var uk) && uk is string uks)
                    {
                        ukMode = uks switch
                        {
                            "strip" => UnknownKeyMode.Strip,
                            "allow" => UnknownKeyMode.Allow,
                            _ => UnknownKeyMode.Reject,
                        };
                    }
                    schema = new ObjectSchema(shape, ukMode);
                    break;
                }
                case "record":
                {
                    // Corpus uses "values", export uses "valueSchema"
                    var valueNode = node.TryGetValue("values", out var v) ? v
                        : node.TryGetValue("valueSchema", out var vs) ? vs : null;
                    schema = new RecordSchema(ImportNode(valueNode));
                    break;
                }
                case "union":
                {
                    var variants = node.TryGetValue("variants", out var v) && v is List<object?> list
                        ? list.Select(ImportNode).ToArray()
                        : Array.Empty<Schema>();
                    schema = new UnionSchema(variants);
                    break;
                }
                case "intersection":
                {
                    var allOf = node.TryGetValue("allOf", out var a) && a is List<object?> list
                        ? list.Select(ImportNode).ToArray()
                        : Array.Empty<Schema>();
                    schema = new IntersectionSchema(allOf);
                    break;
                }
                case "optional":
                {
                    var innerNode = node.TryGetValue("schema", out var s) ? s
                        : node.TryGetValue("inner", out var i) ? i : null;
                    schema = new OptionalSchema(ImportNode(innerNode));
                    break;
                }
                case "nullable":
                {
                    var innerNode = node.TryGetValue("schema", out var s) ? s
                        : node.TryGetValue("inner", out var i) ? i : null;
                    schema = new NullableSchema(ImportNode(innerNode));
                    break;
                }
                case "ref":
                {
                    var refPath = node.TryGetValue("ref", out var r) ? r?.ToString() ?? "" : "";
                    var defName = refPath.Replace("#/definitions/", "");
                    schema = new RefSchema(refPath, () =>
                    {
                        if (resolvedDefs.TryGetValue(defName, out var cached))
                            return cached;
                        if (!definitions.TryGetValue(defName, out var defNode))
                            throw new InvalidOperationException($"Unresolved definition: {defName}");
                        var resolved = ImportNode(defNode);
                        resolvedDefs[defName] = resolved;
                        return resolved;
                    });
                    break;
                }
                default:
                    throw new InvalidOperationException($"Unsupported schema kind: {kind}");
            }

            // Apply default
            if (node.TryGetValue("default", out var defaultVal) && defaultVal is not null)
            {
                schema = schema.Default(defaultVal);
            }

            // Apply coercion
            if (node.TryGetValue("coerce", out var coerceVal) && coerceVal is not null)
            {
                var config = Coercion.NormalizeConfig(coerceVal);
                schema = schema.Coerce(config);
            }

            return schema;
        }

        return ImportNode(doc.Root);
    }

    private static NumberSchema ApplyNumericConstraints(NumberSchema schema, Dictionary<string, object?> node)
    {
        var s = schema;
        if (node.TryGetValue("min", out var min) && min is not null)
            s = s.Min(ToDouble(min));
        if (node.TryGetValue("max", out var max) && max is not null)
            s = s.Max(ToDouble(max));
        if (node.TryGetValue("exclusiveMin", out var emin) && emin is not null)
            s = s.ExclusiveMin(ToDouble(emin));
        if (node.TryGetValue("exclusiveMax", out var emax) && emax is not null)
            s = s.ExclusiveMax(ToDouble(emax));
        if (node.TryGetValue("multipleOf", out var mo) && mo is not null)
            s = s.MultipleOf(ToDouble(mo));
        return s;
    }

    private static Dictionary<string, object?> ToDict(object? obj)
    {
        return obj as Dictionary<string, object?> ?? new Dictionary<string, object?>();
    }

    private static int ToInt(object? value)
    {
        return value switch
        {
            int i => i,
            long l => (int)l,
            double d => (int)d,
            string s => int.Parse(s),
            _ => 0
        };
    }

    private static double ToDouble(object? value)
    {
        return value switch
        {
            double d => d,
            long l => l,
            int i => i,
            float f => f,
            string s => double.Parse(s),
            _ => 0
        };
    }
}
